### ISSUE TITLE
sql/catalog: build composable and well-defined catalog abstractions

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#64089]
Currently in cockroach the access to the catalog (mapping names and IDs to their meanings) is scattered around the code-base. There are a number of layers to this access which are confusing and hard to use. Access is generally a session-level concept, though there are a variety of needs to access the catalog in asynchronous and background jobs. In order to make [Transactional Schema Changes](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20200909_transactional_schema_changes.md) work we'll need to be able to properly orchestrate the transaction used for lookups (#56588) as well as controlling deadlines. All of this hints at a need for more abstraction both above and below the `*descs.Collection`. 

Tons of resolution functionality exists hanging off the `*sql.planner` as it implements a variety of resolution-related interfaces.
This is a meta-issue for tracking a relatively large set of work related to these catalog improvements. Other work to improve the testability and reliability will be performed along the way.

* [ ] Provide a mechanism to inject an abstraction for use by the descs.Collection for precise physical lookup of descriptors.
* [ ] Ensure that all descriptor resolution interfaces (precise, imprecise, glob) are implemented outside of the `sql` package.
* [ ] Ensure all descriptor access utilizes these interfaces; eliminate the possibility of buggy, low-level access.

**Comment by postamar** [comment]
@ajwerner correct me if I'm wrong but at this point, this boils down to:
- getting rid of the internalLookupCtx
- getting rid of the Accessor interface
- refactoring the Resolver interface

right?

**Comment by ajwerner** [comment]
Yes. With the resolver interface move would be a moving of privilege checking too I believe.
